lib/Dialect/TritonGPU/Transforms/Pipeliner/SoftwarePipeliner.cpp

```diff
@@ -71,8 +76,26 @@ struct PipelinePass : public TritonGPUPipelineBase<PipelinePass> {
     if (this->numStages <= 1)
       return;
     SmallVector<scf::ForOp> loops;
-    getOperation()->walk([&](scf::ForOp forOp) { loops.push_back(forOp); });
+
+    getOperation()->walk<WalkOrder::PostOrder>(
+        [&loops](scf::ForOp forOp) { loops.push_back(forOp); });
```

Comment:
Maybe it could be simplified?

Once you find a forOp, you call [getParentOfType](https://mlir.llvm.org/doxygen/classmlir_1_1Operation.html#a9968b118b25a099be92400e88239153c) to go up and look for a parent forOp and remove it from loops?

Reply:
@ThomasRaoux in his previous comment suggested that it's maybe better not to assume the order in which `walk()` traverses operations. What you suggest would probably work though with `WalkOrder::PreOrder`, I don't have a strong opinion on this.

Reply:
OK, it's fine to keep it unchanged.

Just quick check, if you walk preorder, shouldn't the parent of a forOp be visited before the forOp itself?

Reply:
The internet says so. I always have to look up which is which. I believe that's why Thomas suggested against depending on it, so whoever furiously searches for bug around this area has one less thing to double check. I will remove the template to avoid confusion.